### PR TITLE
DOC: create folders in to_csv #24306

### LINKED ISSUE #24306: to_csv  - allow to create folders.
currently `to_csv ` will create a file if it does not exist. However it will not create a folder (nor sub-folders).
In this case it will raise 
`IOError: [Errno 2] No such file or directory:`

The user must wrap his code in order for it to create a folder.

I'm not sure why the function behaves like this.. It would be great if it will also create folders if needed.

Reference from stack overflow:
https://stackoverflow.com/questions/47143836/pandas-dataframe-to-csv-raising-ioerror-no-such-file-or-directory

### PR DESCRIPTION
**Pull request by feefladder**
- [x] closes  #24306
- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them

### CODE REVIEW

**Comment by Joeperdefloep** [comment]
one question would be if we also want this in `to_excel`

**Comment by mzeitlin11** [comment]
Thanks for the pr @Joeperdefloep! Based on the linked issue, it seems like the idea was to add a doc note, and there was no consensus about adding this feature. I'd be surprised by automatic directory creation being the default just based on python behavior, eg 
```
>>> with open("no_dir/text.txt", "x") as handle:
...     handle.write("hello")
...
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
FileNotFoundError: [Errno 2] No such file or directory: 'no_dir/text.txt'
```

**Comment by Joeperdefloep** [comment]
Hi @mzeitlin11 thanks for your comment. I know that it would not eb default Python behaviour. My proposed PR is also only an upate in the docs on how to create a new folder so that users can wrap it in a function. #42255 does indeed implement this behaviour and there is discussion there on whether this should be implemented and/or exposed.

**Comment by mzeitlin11** [comment]
> Hi @mzeitlin11 thanks for your comment. I know that it would not eb default Python behaviour. My proposed PR is also only an upate in the docs on how to create a new folder so that users can wrap it in a function. #42255 does indeed implement this behaviour and there is discussion there on whether this should be implemented and/or exposed.

My bad @Joeperdefloep, commented on the wrong pr :)

**Comment by Joeperdefloep** [comment]
yeah, actually also slightly mine, should've said *take* on the issue

File: pandas/core/generic.py [hunk 1]
```diff
@@ -3450,6 +3450,17 @@ def to_csv(
...
+        Create 'out.csv' in a new folder 'folder/subfolder/'
+
+        >>> import os
+        >>> os.mkdir('folder/subfolder', parents=True, exists_ok=True)
+        >>> df.to_csv('out.csv')
+
+        >>> from pathlib import Path
+        >>> filepath = Path('folder/subfolder/out.csv')
+        >>> filepath.parent.mkdir(parents=True, exist_ok=True)
+        >>> df_to_csv(filepath)
```
Comment (alimcmaster1) [inline pandas/core/generic.py:3463]: Should be df.to_csv

File: pandas/core/generic.py [hunk 2]
```diff
@@ -3450,6 +3450,17 @@ def to_csv(
         ...                         archive_name='out.csv')  # doctest: +SKIP
         >>> df.to_csv('out.zip', index=False,
         ...           compression=compression_opts)  # doctest: +SKIP
+
+        Create 'out.csv' in a new folder 'folder/subfolder/'
```
Comment (alimcmaster1) [inline pandas/core/generic.py:3454]: Can you please make this a proper sentence. Some like to ```To write a csv file to a new folder you will first need to create a folder using either os or pathlib. ```

File: pandas/core/generic.py [hunk 3]
```diff
@@ -3450,6 +3450,17 @@ def to_csv(
...
+
+        >>> import os
+        >>> os.mkdir('folder/subfolder', parents=True, exists_ok=True)
```
Comment (alimcmaster1) [inline pandas/core/generic.py:3457]: os.mkdir doesn't support `parents` or `exists_ok` please try running the code before submitting a PR. 

https://docs.python.org/3/library/os.html?highlight=os%20mkdir#os.mkdir

File: pandas/core/generic.py [hunk 4]
```diff
@@ -3450,6 +3450,17 @@ def to_csv(
...
+        >>> df.to_csv('out.csv')
+
+        >>> from pathlib import Path
+        >>> filepath = Path('folder/subfolder/out.csv')
```
Comment (alimcmaster1) [inline pandas/core/generic.py:3461]: Do we really need a folder and a subfolder to illustrate the example? Can it just be `folder/out.csv`

**Review by alimcmaster1 — CHANGES_REQUESTED** [review]
Thanks for the PR. Few comments here :)

File: pandas/core/generic.py (hunk 4 above)
Comment (Joeperdefloep) [reply pandas/core/generic.py:3461]: Could be, yes, but then users would be puzzled when having to use nested folders. This method also works for just `folder/out.csv` and with any level of sub-foldering

File: pandas/core/generic.py (hunk 3 above)
Comment (Joeperdefloep) [reply pandas/core/generic.py:3457]: Oops, I copied it wrongly from [here](https://stackoverflow.com/a/11101867/14681457).

File: pandas/core/generic.py (hunk 4 above)
Comment (alimcmaster1) [reply pandas/core/generic.py:3460]: Please don't link python 2 docs, its unsupported by Python.org and by pandas

File: pandas/core/generic.py (hunk 3 above)
Comment (alimcmaster1) [reply pandas/core/generic.py]: Please check code works before committing :)

**Review by alimcmaster1 — CHANGES_REQUESTED** [review]
Few comments and doc build is failing please could you take a look @Joeperdefloep

**Comment by swt2c** [comment]
Why are you adding an empty file called `file.xlsx`?

File: pandas/core/generic.py [hunk 5]
```diff
@@ -3450,6 +3450,18 @@ def to_csv(
         ...                         archive_name='out.csv')  # doctest: +SKIP
         >>> df.to_csv('out.zip', index=False,
         ...           compression=compression_opts)  # doctest: +SKIP
+
```
Comment (jreback) [inline pandas/core/generic.py:3453]: good point, pls delete file.xlsx

**Review by jreback — CHANGES_REQUESTED** [review]

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @Joeperdefloep lgtm.

@alimcmaster1 @lithomas1 if any comments.

**Comment by jreback** [comment]
thanks @Joeperdefloep

**Pull request merged**